### PR TITLE
Feat/improve send form

### LINKED ISSUE #1040: No error message when sending the entire balance from the wallet and not accounting for the fee
No error message when sending the entire balance from the wallet and not accounting for the fee. If I send more than my current balance the error message appears but if I send exactly my balance not including fees it allows me to do so.

The user cannot send there entire balance because we must subtract the fees from the total first. The desktop wallet handles this well we can use that as an example.

Screen Shot 2021-03-01 at 7 57 04 AM
Screen Shot 2021-03-01 at 7 56 57 AM

### PR DESCRIPTION
**Pull request by kyranjamie**
> Try out this version of the Stacks Wallet - download [extension builds](https://github.com/blockstack/ux/actions/runs/963905400).

Closes #1040

Also default selects the asset you have when you only have 1️⃣ 

@aulneau I've renamed they keys to be more descriptive & use enum convention (PascalCase) in 28f7b99

- [x] Display error when sending > balance, minus fee

### CODE REVIEW

File: src/components/asset-search/asset-search.tsx
Comment (fbwoolf) [inline src/components/asset-search/asset-search.tsx]: Prob mean to remove this?

File: src/pages/popup/send-form.tsx
Comment (fbwoolf) [inline src/pages/popup/send-form.tsx]: Just noting these if you want to remove?

File: src/pages/popup/send-form.tsx
Comment (fbwoolf) [inline src/pages/popup/send-form.tsx]: What is this?

**Review by fbwoolf — APPROVED** [review]
Looks good!

File: .changeset/neat-cameras-share.md
Comment (aulneau) [inline .changeset/neat-cameras-share.md]: I'm hoping we can start using a bit more verbose updates here. I think the more time we spend here, the better the experience will be for folks who come to check out what changed.

File: src/components/send/amount-field.tsx
Comment (aulneau) [inline src/components/send/amount-field.tsx:1]: This file was not actually deleted, but due to the way it was moved, git thinks it was. Can you make it so the history is preserved?

File: src/components/asset-search/asset-search.tsx
Comment (aulneau) [inline src/components/asset-search/asset-search.tsx]: what does the nullish operator do in this context?

File: src/common/hooks/use-send-form.ts
Comment (aulneau) [inline src/common/hooks/use-send-form.ts]: Would you be willing to include changes in this PR to take into account locked balance?

**Review by aulneau — APPROVED** [review]

File: .changeset/neat-cameras-share.md
Comment (kyranjamie) [reply .changeset/neat-cameras-share.md]: Can you suggest a friendlier alternative?

File: src/store/transactions/requests.ts
Comment (aulneau) [inline src/store/transactions/requests.ts]: I wonder if this should actually not go in this file. This file is meant more to hold any atoms related to the request token (jwt). This fee selector seems to be more like an api request.

Comment (kyranjamie) [reply src/store/transactions/requests.ts]: I've created an API req file containing this. 

Is there any reason to put the API clients in the store, versus exposing them via hooks (`useMemo` with a dep on network?).

Comment (aulneau) [reply src/store/transactions/requests.ts]: Yes because you cannot use hooks in atoms and often we need a statefully aware client in other atoms. It would fine to use these client atoms to create hooks, but not the other way around.

Comment (aulneau) [reply src/store/transactions/requests.ts]: > Is there any reason to put the API clients in the store, versus exposing them via hooks (useMemo with a dep on network?).

Also, we should ideally always have atoms be the place where fetching of data happens, and not directly in hooks (if that was what you were suggesting).

Comment (kyranjamie) [reply src/store/transactions/requests.ts]: Which code is responsible for cache invalidation/refetching etc, all the features of swr/react-query? Can we use any of these libraries or do we have to implement it ourselves ?

**Comment by aulneau** [comment]
@kyranjamie can you rebase this PR? tests should pass if you do

File: src/store/transactions/requests.ts
Comment (aulneau) [reply src/store/transactions/requests.ts]: > Which code is responsible for cache invalidation/refetching etc, all the features of swr/react-query? Can we use any of these libraries or do we have to implement it ourselves ?

Not sure what you're getting at, can you rephrase?

We can make use of react-query, but we should dig into that after the jotai migration.

Comment (kyranjamie) [reply src/store/transactions/requests.ts]: These libraries bring a lot of functionality, such as cache invalidation and methods to invoke refetching of data. 

We don't have these libraries, so wondering where, in this apps' architecture, that functionality would reside.

Comment (aulneau) [reply src/store/transactions/requests.ts]: ah okay yes, makes sense! I 100% want to use react-query as I would hate to recreate and maintain all that functionality. The amazing thing about jotai is you can leverage all of the functionality of react-query but in atom state. [see here in heystack](https://github.com/blockstack/heystack/blob/main/src/store/hey.ts#L27) on how that's used. This is better than relying on using hooks for react-query because we can't then use the data they fetch in store without doing some higher level syncing of their fetches and putting it back into atom state (eg react-query hook -> onComplete -> setAtom(data)). With the use of atomWithQuery, we get all the stuff we love, and other atoms can rely directly on those query atoms and the data they fetch, and when they update, all the atoms that depend on it update too!

**Comment by aulneau** [comment]
This seems to no longer be working, see video:


https://user-images.githubusercontent.com/11803153/122064957-2c3f7e80-cdb7-11eb-9171-7b9141b5070c.mp4

**Comment by aulneau** [comment]
@kyranjamie yeah something feels a lot different about these changes

- when selecting send for the first time, there is a long delay that does not exist on `main`
- can't select STX
- when a different token is selected, can't switch
- generally feels slower for some reason 🤔

**Comment by aulneau** [comment]
Here is side by side:

### main
https://user-images.githubusercontent.com/11803153/122066266-39a93880-cdb8-11eb-911f-dae4eeec2563.mp4

### this branch
https://user-images.githubusercontent.com/11803153/122066270-3a41cf00-cdb8-11eb-90ee-1d358f39673b.mp4

**Review by aulneau — DISMISSED** [review]
see https://github.com/blockstack/stacks-wallet-web/pull/1285#issuecomment-861521778

**Comment by kyranjamie** [comment]
Thanks @aulneau—investigating now

I've added a render prop component that does the fee fetching so I can add a suspense boundary just for the send max button.

```jsx
  <Suspense fallback={<SendMaxButton isDisabled />}>
    {balances.value && selectedAsset ? (
      <FeeFetcher>
        {fee => <SendMaxButton onClick={() => handleSetSendMax(fee)} />}
      </FeeFetcher>
    ) : null}
  </Suspense>
```

This speeds up load (running in 2g throttle).

I can't get replicate the bugs selecting assets, though.

**Comment by andresgalante** [comment]
Let's wait to merge this one after #1304

**Comment by aulneau** [comment]
it works!! none of the bugs I saw before are present, great job!

I'd recommend dropping this commit https://github.com/blockstack/stacks-wallet-web/pull/1285/commits/a4ea6ea08f18ace9b6d27412f9f2944fd70221fe as it will likely create loads of conflicts once the jotai PR merges.

**Comment by kyranjamie** [comment]
Fixed merge issues and tested, will merge

**Pull request merged**
